Allow observers to set `billingAddress` by returning `billingData`

opr opened the pull request:
This is required since we didn't correctly deprecate `billingData` when we changed the name to `billingAddress`, so some extensions could still be returning `billingData` to the `onPaymentProcessing` observer.

<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [ ] E2E tests

#### Internal Developer Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Install the Stripe plugin from source.
2. Go to https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8ffd22aff3b06eda02a1ae2fd8368b71450b36a9/client/blocks/credit-card/use-payment-processing.js#L136 and change `billingData` to:
```js
billingData: {
  ...bilingData,
  first_name: 'Set from',
  last_name: 'Stripe plugin',
}
```
3. Add `shippingData` too - this can be a copy of `billingData`.
4. Add items to your cart and load the Checkout block.
5. Open the console and enable `Preserve log`.
6. Check out using Stripe.
7. Ensure you see a deprecated notice for `billingData`.
8. Edit the previous code, and change the key: `billingData` to `billingAddress`
```js
billingAddress: {
  ...bilingData,
  first_name: 'Set from',
  last_name: 'Stripe plugin',
}
```
9. Also add `shippingAddress` - this can be a copy of `billingAddress`.
9. Check out again and ensure the deprecation notice does not appear.

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Skipping.

opr committed:
Allow observers to set billingAddress by returning billingData

This is required since we didn't correctly deprecate billingData when we changed the name to billingAddress

opr committed:
Add link to original PR

alexflorisca reviewed (changes requested):
Nice catch, we also need to do this in the errorResponse condition as well. We could maybe optimise it so we destructure the data from `response.meta` outside the `if...else` statement

opr committed:
Set billingAddress when observer errors

alexflorisca reviewed (approved):
Noice one, looks good 👍

opr committed:
Rename  shippingData to shippingAddress

opr committed:
Add isBillingAddress and isShippingAddress type guards

opr committed:
Add tests for new type guards

opr committed:
Only set billing and shipping if they are valid objects

opr committed:
Add tests for __internalEmitPaymentProcessingEvent thunk

opr committed:
Update deprecated version

opr committed:
Return promise from this function to aid with testability

opr committed:
Add tests for shippingAddress and paymentMethodData

opr committed:
Ensure correct value is used to set shipping address

opr committed:
Move test data out of tests to aid with reusability

opr committed:
Improve success callback name

opr committed:
Add mocked __internalSetPaymentMethodData to correct object

It was in registry, but should be in dispatch as the action is on the same store as the thunk. Registry is used for actions on other stores.

opr committed:
Add test for failed observers

opr committed:
Add test for mixed observers

opr commented:
@alexflorisca I added tests, and also changed `shippingData` to `shippingAddress` - please re-test, and also re-test with `shippingData` and `shippingAddress`, as well as with `billingData`.

alexflorisca commented on assets/js/data/payment/thunks.ts:
```diff
@@ -64,25 +72,57 @@ export const __internalEmitPaymentProcessingEvent: emitProcessingEventType = (
 				) {
 					errorResponse = response;
 				}
+				const {
```
It would be helpful to add a comment here to explain this craziness

opr replied:
Good idea, added.

alexflorisca reviewed (approved):
Nice one, tested with the shippingAddress and all good. Thanks for writing the tests and the type-guards as well, stellar work ⭐

opr committed:
Add comments explaining the destructure & deprecation

opr commented:
I think the E2E fails are unrelated to these changes. I will merge this.
